orc8r/gateway/python/magma/ctraced/trace_manager.py

```diff
@@ -0,0 +1,104 @@
+"""
+Copyright 2020 The Magma Authors.
+
+This source code is licensed under the BSD-style license found in the
+LICENSE file in the root directory of this source tree.
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import logging
+import os
+import pathlib
+import subprocess
+
+from subprocess import SubprocessError
+
+TRACE_FILE_NAME = "call_trace.pcap"
+
+class TraceManager:
+    """
+    TraceManager is a wrapper for tshark specifically for starting and stopping
+    call/interface/subscriber traces.
+
+    Only a single trace can be captured at a time.
+    """
+    def __init__(self, config):
+        self._is_active = False # is call trace being captured
+        self._proc = None
+        self._file_name = None
+        self._trace_directory = config.get("trace_directory",
+                                           "/var/opt/magma/trace")
+        self._trace_interface = config.get("trace_interface", "eth0")
+
+
+    def start_trace(self) -> bool:
+        """
+        Start a call trace.
+        Captures all packets across the eth0 interface.
+        TODO (andreilee): Make configurable for trace type
+        :return: True if successful
+        """
+        if self._is_active:
+            logging.error("Failed to start trace: Trace already active")
+            return False
+        command = ["tshark"]
+
+        # Specify southbound interfaces
+        command.extend(["-i", self._trace_interface])
+
+        # Limit to ~4 MiB
+        # TODO (andreilee): Make configurable
+        command.extend(["-a", "filesize:4000"])
+
+        # Specify output file
+        self._file_name = self._build_trace_filename()
+        command.extend(["-w", self._file_name])
+
+        logging.info("Starting trace with tshark, command: [%s]",
+                     ' '.join(command))
+
+        self._ensure_trace_directory_exists()
+
+        try:
+            self._proc = subprocess.Popen(command)
+        except SubprocessError as e:
+            logging.error("Failed to start trace: %s", str(e))
+            return False
+
+        self._is_active = True
+        logging.info("Successfully started trace with tshark")
+        return True
+
+    def end_trace(self) -> bytes:
+        """
+        :return: Call trace file in bytes
+        """
+        # If trace is active, then stop it
+        self._proc.poll()
```

Comment:
If we crash and restart, self._proc might be set to None,  check for _proc prior to calling poll()

Reply:
shouldnt the manager use `self._is_active` to check if is running?